File: hello-websocket-swift/common/Codec.swift
// Hello WebSocket Protocol Codec - Swift implementation.
//
// Implements the canonical binary protocol defined in PROTOCOL.md.
// Provides constants, primitive encoders/decoders, message types, frame codec,
// SHA-256 hash, and message dispatch for all 13 message types.

import Foundation
import ucrt

// MARK: - SHA-256 Implementation

struct SHA256 {
    private static let k: [UInt32] = [
        0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5, 0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
        0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3, 0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
        0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc, 0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
        0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7, 0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
        0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13, 0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
        0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
        0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
        0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208, 0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2
    ]

    static func hash(_ data: [UInt8]) -> [UInt8] {
        var h: [UInt32] = [
            0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a,
            0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19
        ]

        var msg = data
        let bitLen = UInt64(msg.count) * 8

        // Padding
        msg.append(0x80)
        while msg.count % 64 != 56 {
            msg.append(0)
        }
        for i in (0..<8).reversed() {
            msg.append(UInt8((bitLen >> (i * 8)) & 0xFF))
        }

        // Process each 512-bit block
        for blockStart in stride(from: 0, to: msg.count, by: 64) {
            var w = [UInt32](repeating: 0, count: 64)
            for i in 0..<16 {
                let b0 = UInt32(msg[blockStart + i * 4])
                let b1 = UInt32(msg[blockStart + i * 4 + 1])
                let b2 = UInt32(msg[blockStart + i * 4 + 2])
                let b3 = UInt32(msg[blockStart + i * 4 + 3])
                w[i] = (b0 << 24) | (b1 << 16) | (b2 << 8) | b3
            }
            for i in 16..<64 {
                let s0 = rotr(w[i-15], 7) ^ rotr(w[i-15], 18) ^ (w[i-15] >> 3)
                let s1 = rotr(w[i-2], 17) ^ rotr(w[i-2], 19) ^ (w[i-2] >> 10)
                w[i] = w[i-16] &+ s0 &+ w[i-7] &+ s1
            }

            var a = h[0], b = h[1], c = h[2], d = h[3]
            var e = h[4], f = h[5], g = h[6], hh = h[7]

            for i in 0..<64 {
                let s1 = rotr(e, 6) ^ rotr(e, 11) ^ rotr(e, 25)
                let ch = (e & f) ^ (~e & g)
                let t1 = hh &+ s1 &+ ch &+ k[i] &+ w[i]
                let s0 = rotr(a, 2) ^ rotr(a, 13) ^ rotr(a, 22)
                let maj = (a & b) ^ (a & c) ^ (b & c)
                let t2 = s0 &+ maj
                hh = g; g = f; f = e; e = d &+ t1
                d = c; c = b; b = a; a = t1 &+ t2
            }

            h[0] = h[0] &+ a; h[1] = h[1] &+ b; h[2] = h[2] &+ c; h[3] = h[3] &+ d
            h[4] = h[4] &+ e; h[5] = h[5] &+ f; h[6] = h[6] &+ g; h[7] = h[7] &+ hh
        }

        var result = [UInt8]()
        for v in h {
            result.append(UInt8((v >> 24) & 0xFF))
            result.append(UInt8((v >> 16) & 0xFF))
            result.append(UInt8((v >> 8) & 0xFF))
            result.append(UInt8(v & 0xFF))
        }
        return result
    }

    private static func rotr(_ x: UInt32, _ n: UInt32) -> UInt32 {
        return (x >> n) | (x << (32 - n))
    }
}

// MARK: - Constants

public enum WsProtocol {
    public static let port: Int = 9898
    public static let magic: UInt8 = 0x48
    public static let version: UInt8 = 0x01
    public static let headerLen: Int = 8
    public static let serverLang = "SWIFT"
    public static let clientLang = "SWIFT"
}

// Message types
public enum MsgType: UInt8 {
    case hello = 0x01
    case bonjour = 0x02
    case echoRequest = 0x03
    case echoResponse = 0x04
    case kissRequest = 0x05
    case kissResponse = 0x06
    case ping = 0x07
    case pong = 0x08
    case timeNotification = 0x09
    case randomNumber = 0x0A
    case hashResponse = 0x0B
    case disconnect = 0x0C
    case error = 0x7F
}

// Error codes
public enum ErrCode: Int32 {
    case decode = 0x01
    case unknownMsgType = 0x02
    case truncatedPayload = 0x03
    case badMagic = 0x04
    case badVersion = 0x05
    case sessionNotFound = 0x06
    case internal_ = 0x07
}

// MARK: - Byte Writer

public final class ByteWriter {
    private var buf: [UInt8] = []

    public init() {}

    public func writeU8(_ v: UInt8) { buf.append(v) }
    public func writeU16(_ v: UInt16) {
        buf.append(UInt8((v >> 8) & 0xFF))
        buf.append(UInt8(v & 0xFF))
    }
    public func writeU32(_ v: UInt32) {
        buf.append(UInt8((v >> 24) & 0xFF))
        buf.append(UInt8((v >> 16) & 0xFF))
        buf.append(UInt8((v >> 8) & 0xFF))
        buf.append(UInt8(v & 0xFF))
    }
    public func writeI32(_ v: Int32) { writeU32(UInt32(bitPattern: v)) }
    public func writeI64(_ v: Int64) { writeU64(UInt64(bitPattern: v)) }
    public func writeU64(_ v: UInt64) {
        for i in (0..<8).reversed() {
            buf.append(UInt8((v >> (i * 8)) & 0xFF))
        }
    }
    public func writeString(_ s: String) {
        let b = Array(s.utf8)
        writeU32(UInt32(b.count))
        buf.append(contentsOf: b)
    }
    public func writeKv(_ m: [(String, String)]) {
        writeU32(UInt32(m.count))
        for (k, v) in m {
            writeString(k)
            writeString(v)
        }
    }
    public func data() -> [UInt8] { return buf }
}

// MARK: - Byte Reader

public final class ByteReader {
    private let data: [UInt8]
    private var pos: Int = 0

    public init(_ data: [UInt8]) { self.data = data }
    public init(_ data: Data) { self.data = Array(data) }

    public func readU8() throws -> UInt8 {
        guard pos + 1 <= data.count else { throw CodecError.eof("u8 at \(pos)") }
        let v = data[pos]; pos += 1; return v
    }
    public func readU16() throws -> UInt16 {
        guard pos + 2 <= data.count else { throw CodecError.eof("u16 at \(pos)") }
        let v = (UInt16(data[pos]) << 8) | UInt16(data[pos+1])
        pos += 2; return v
    }
    public func readU32() throws -> UInt32 {
        guard pos + 4 <= data.count else { throw CodecError.eof("u32 at \(pos)") }
        var v: UInt32 = 0
        for i in 0..<4 { v = (v << 8) | UInt32(data[pos+i]) }
        pos += 4; return v
    }
    public func readI32() throws -> Int32 { Int32(bitPattern: try readU32()) }
    public func readI64() throws -> Int64 { Int64(bitPattern: try readU64()) }
    public func readU64() throws -> UInt64 {
        guard pos + 8 <= data.count else { throw CodecError.eof("i64 at \(pos)") }
        var v: UInt64 = 0
        for i in 0..<8 { v = (v << 8) | UInt64(data[pos+i]) }
        pos += 8; return v
    }
    public func readString() throws -> String {
        let ln = Int(try readU32())
        guard pos + ln <= data.count else { throw CodecError.eof("string at \(pos), len=\(ln)") }
        let s = String(decoding: data[pos..<pos+ln], as: UTF8.self)
        pos += ln; return s
    }
    public func readKv() throws -> [(String, String)] {
        let count = Int(try readU32())
        var m: [(String, String)] = []
        for _ in 0..<count {
            let k = try readString()
            let v = try readString()
            m.append((k, v))
        }
        return m
    }
}

// MARK: - Errors

public enum CodecError: Error, CustomStringConvertible {
    case eof(String)
    case frameTooShort(Int)
    case badMagic(UInt8)
    case badVersion(UInt8)
    case truncatedPayload(declared: Int, available: Int)
    case unknownMsgType(UInt8)

    public var description: String {
        switch self {
        case .eof(let s): return "unexpected end of data reading \(s)"
        case .frameTooShort(let n): return "frame too short: \(n) bytes"
        case .badMagic(let v): return String(format: "bad magic: 0x%02x", v)
        case .badVersion(let v): return String(format: "bad version: 0x%02x", v)
        case .truncatedPayload(let d, let a): return "truncated payload: declared \(d), available \(a)"
        case .unknownMsgType(let v): return String(format: "unknown message type: 0x%02x", v)
        }
    }
}

// MARK: - Frame Codec

public func encodeFrame(_ msgType: MsgType, payload: [UInt8]) -> [UInt8] {
    var buf = [UInt8]()
    buf.append(WsProtocol.magic)
    buf.append(WsProtocol.version)
    buf.append(msgType.rawValue)
    buf.append(0x00) // flags
    let len = UInt32(payload.count)
    buf.append(UInt8((len >> 24) & 0xFF))
    buf.append(UInt8((len >> 16) & 0xFF))
    buf.append(UInt8((len >> 8) & 0xFF))
    buf.append(UInt8(len & 0xFF))
    buf.append(contentsOf: payload)
    return buf
}

public func decodeFrame(_ data: [UInt8]) throws -> (MsgType, [UInt8]) {
    guard data.count >= WsProtocol.headerLen else {
        throw CodecError.frameTooShort(data.count)
    }
    guard data[0] == WsProtocol.magic else {
        throw CodecError.badMagic(data[0])
    }
    guard data[1] == WsProtocol.version else {
        throw CodecError.badVersion(data[1])
    }
    guard let msgType = MsgType(rawValue: data[2]) else {
        throw CodecError.unknownMsgType(data[2])
    }
    let payloadLen = (UInt32(data[4]) << 24) | (UInt32(data[5]) << 16) | (UInt32(data[6]) << 8) | UInt32(data[7])
    let avail = data.count - WsProtocol.headerLen
    guard Int(payloadLen) <= avail else {
        throw CodecError.truncatedPayload(declared: Int(payloadLen), available: avail)
    }
    let payload = Array(data[WsProtocol.headerLen..<WsProtocol.headerLen+Int(payloadLen)])
    return (msgType, payload)
}

// MARK: - Message

public struct WsMessage {
    public var type: MsgType
    public var clientLanguage: String?
    public var serverLanguage: String?
    public var echoId: Int64?
    public var echoMeta: String?
    public var echoData: String?
    public var echoStatus: Int32?
    public var echoResults: [(idx: Int64, type: UInt8, kv: [(String, String)])]?
    public var osName: String?
    public var osVersion: String?
    public var osRelease: String?
    public var osArch: String?
    public var kissLanguage: String?
    public var kissEncoding: String?
    public var kissTimeZone: String?
    public var timestampMs: Int64?
    public var iso8601: String?
    public var randomId: Int64?
    public var randomNumber: Int64?
    public var hashHex: String?
    public var disconnectReason: String?
    public var errorCode: Int32?
    public var errorMessage: String?

    public init(type: MsgType) { self.type = type }
}

public func decodeMessage(_ data: [UInt8]) throws -> WsMessage {
    let (msgType, payload) = try decodeFrame(data)
    let r = ByteReader(payload)
    var msg = WsMessage(type: msgType)

    switch msgType {
    case .hello:
        msg.clientLanguage = try r.readString()
    case .bonjour:
        msg.serverLanguage = try r.readString()
    case .echoRequest:
        msg.echoId = try r.readI64()
        msg.echoMeta = try r.readString()
        msg.echoData = try r.readString()
    case .echoResponse:
        msg.echoStatus = try r.readI32()
        let count = Int(try r.readU32())
        var results: [(idx: Int64, type: UInt8, kv: [(String, String)])] = []
        for _ in 0..<count {
            let idx = try r.readI64()
            let typ = try r.readU8()
            let kv = try r.readKv()
            results.append((idx, typ, kv))
        }
        msg.echoResults = results
    case .kissRequest:
        msg.osName = try r.readString()
        msg.osVersion = try r.readString()
        msg.osRelease = try r.readString()
        msg.osArch = try r.readString()
    case .kissResponse:
        msg.kissLanguage = try r.readString()
        msg.kissEncoding = try r.readString()
        msg.kissTimeZone = try r.readString()
    case .ping:
        msg.timestampMs = try r.readI64()
    case .pong:
        msg.timestampMs = try r.readI64()
    case .timeNotification:
        msg.timestampMs = try r.readI64()
        msg.iso8601 = try r.readString()
    case .randomNumber:
        msg.randomId = try r.readI64()
        msg.randomNumber = try r.readI64()
    case .hashResponse:
        msg.randomId = try r.readI64()
        msg.hashHex = try r.readString()
    case .disconnect:
        msg.disconnectReason = try r.readString()
    case .error:
        msg.errorCode = try r.readI32()
        msg.errorMessage = try r.readString()
    }

    return msg
}

// MARK: - Utility

public func nowMs() -> Int64 {
    return Int64(Date().timeIntervalSince1970 * 1000)
}

public func nowISO() -> String {
    let ts = Date()
    let secs = Int64(ts.timeIntervalSince1970)
    let dayNum = Int(secs / 86400)
    let rem = secs % 86400
    let h = Int(rem / 3600)
    let m = Int((rem / 60) % 60)
    let s = Int(rem % 60)
    var y = 1970
    var d = dayNum
    while true {
        let leap = (y % 4 == 0 && y % 100 != 0) || y % 400 == 0
        let yearDays = leap ? 366 : 365
        if d < yearDays { break }
        d -= yearDays
        y += 1
    }
    let leap = (y % 4 == 0 && y % 100 != 0) || y % 400 == 0
    let monthDays = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    var month = 1
    for md in monthDays {
        if d < md { break }
        d -= md
        month += 1
    }
    return String(format: "%04d-%02d-%02dT%02d:%02d:%02dZ", y, month, d + 1, h, m, s)
}

public func hashNumber(_ num: Int64) -> String {
    let s = String(num)
    let bytes = Array(s.utf8)
    let fullHash = SHA256.hash(bytes)
    // First 5 bytes = 10 hex chars
    let prefix = fullHash.prefix(5)
    return prefix.map { String(format: "%02x", $0) }.joined()
}

public func logMsg(_ name: String, _ msg: String) {
    let ts = Date()
    let secs = Int64(ts.timeIntervalSince1970)
    let h = (secs / 3600) % 24
    let m = (secs / 60) % 60
    let s = secs % 60
    let dateStr = String(format: "%02d:%02d:%02d", h, m, s)
    print("[\(dateStr)] [INFO] [\(name)] \(msg)")
    fflush(stdout)
}

File: hello-websocket-php/common/Codec.php
<?php
/**
 * Hello WebSocket Protocol Codec - PHP implementation.
 *
 * Implements the canonical binary protocol defined in PROTOCOL.md.
 * Provides constants, primitive encoders/decoders, message types, frame codec,
 * and message dispatch for all 13 message types.
 */

namespace HelloWs\Common;

// ─── Constants ──────────────────────────────────────────────────────────

const PORT = 9898;
const MAGIC = 0x48;
const VERSION = 0x01;
const HEADER_LEN = 8;
const SERVER_LANG = 'PHP';
const CLIENT_LANG = 'PHP';

// Message types
const MSG_HELLO = 0x01;
const MSG_BONJOUR = 0x02;
const MSG_ECHO_REQUEST = 0x03;
const MSG_ECHO_RESPONSE = 0x04;
const MSG_KISS_REQUEST = 0x05;
const MSG_KISS_RESPONSE = 0x06;
const MSG_PING = 0x07;
const MSG_PONG = 0x08;
const MSG_TIME_NOTIFICATION = 0x09;
const MSG_RANDOM_NUMBER = 0x0A;
const MSG_HASH_RESPONSE = 0x0B;
const MSG_DISCONNECT = 0x0C;
const MSG_ERROR = 0x7F;

// Error codes
const ERR_DECODE = 0x01;
const ERR_UNKNOWN_MSG_TYPE = 0x02;
const ERR_TRUNCATED_PAYLOAD = 0x03;
const ERR_BAD_MAGIC = 0x04;
const ERR_BAD_VERSION = 0x05;
const ERR_SESSION_NOT_FOUND = 0x06;
const ERR_INTERNAL = 0x07;

// Intervals (seconds, used as float for sub-second precision in timers)
const PING_INTERVAL = 1.0;
const SESSION_TIMEOUT = 60.0;
const TIME_INTERVAL = 5.0;
const RANDOM_INTERVAL = 5.0;
const KISS_INTERVAL = 5.0;

// ─── Primitive Encoders ──────────────────────────────────────────────────

class ByteWriter
{
    private string $buf = '';

    public function writeU8(int $v): void
    {
        $this->buf .= chr($v & 0xFF);
    }

    public function writeU16(int $v): void
    {
        $this->buf .= pack('n', $v & 0xFFFF);
    }

    public function writeU32(int $v): void
    {
        $this->buf .= pack('N', $v & 0xFFFFFFFF);
    }

    public function writeI32(int $v): void
    {
        $this->buf .= pack('N', $v & 0xFFFFFFFF);
    }

    public function writeI64(int $v): void
    {
        // PHP ints are 64-bit signed; arithmetic right shift preserves sign
        $high = ($v >> 32) & 0xFFFFFFFF;
        $low = $v & 0xFFFFFFFF;
        $this->buf .= pack('NN', $high, $low);
    }

    public function writeString(string $s): void
    {
        $b = $s;
        $this->writeU32(strlen($b));
        $this->buf .= $b;
    }

    /** @param array<string,string> $m */
    public function writeKv(array $m): void
    {
        $this->writeU32(count($m));
        foreach ($m as $k => $v) {
            $this->writeString((string)$k);
            $this->writeString((string)$v);
        }
    }

    public function data(): string
    {
        return $this->buf;
    }
}

// ─── Primitive Decoders ─────────────────────────────────────────────────

class ByteReader
{
    private string $data;
    private int $pos = 0;
    private int $len;

    public function __construct(string $data)
    {
        $this->data = $data;
        $this->len = strlen($data);
    }

    public function remaining(): int { return $this->len - $this->pos; }

    private function requireBytes(int $size, string $field): void
    {
        if ($size < 0 || $size > $this->remaining()) {
            throw new \InvalidArgumentException("$field requires $size bytes, only {$this->remaining()} remain");
        }
    }

    public function readU8(): int
    {
        $this->requireBytes(1, 'u8');
        $v = ord($this->data[$this->pos]);
        $this->pos++;
        return $v;
    }

    public function readU16(): int
    {
        $this->requireBytes(2, 'u16');
        $v = unpack('n', substr($this->data, $this->pos, 2))[1];
        $this->pos += 2;
        return $v;
    }

    public function readU32(): int
    {
        $this->requireBytes(4, 'u32');
        $v = unpack('N', substr($this->data, $this->pos, 4))[1];
        $this->pos += 4;
        return $v;
    }

    public function readI32(): int
    {
        $this->requireBytes(4, 'i32');
        $v = unpack('N', substr($this->data, $this->pos, 4))[1];
        $this->pos += 4;
        // Convert to signed
        if ($v >= 2147483648) {
            $v -= 4294967296;
        }
        return $v;
    }

    public function readI64(): int
    {
        $this->requireBytes(8, 'i64');
        $high = unpack('N', substr($this->data, $this->pos, 4))[1];
        $low = unpack('N', substr($this->data, $this->pos + 4, 4))[1];
        $this->pos += 8;
        // Combine using pack/unpack to avoid float precision loss
        $packed = pack('NN', $high, $low);
        $v = unpack('J', $packed)[1];
        // unpack('J') returns int for values <= PHP_INT_MAX, float otherwise
        if (is_float($v)) {
            $v = (int)($v - 18446744073709551616.0);
        }
        return $v;
    }

    public function readString(): string
    {
        $ln = $this->readU32();
        $this->requireBytes($ln, 'string');
        $s = substr($this->data, $this->pos, $ln);
        $this->pos += $ln;
        return $s;
    }

    /** @return array<string,string> */
    public function readKv(): array
    {
        $count = $this->readU32();
        if ($count > intdiv($this->remaining(), 8)) {
            throw new \InvalidArgumentException("kv count $count exceeds remaining payload");
        }
        $m = [];
        for ($i = 0; $i < $count; $i++) {
            $k = $this->readString();
            $v = $this->readString();
            $m[$k] = $v;
        }
        return $m;
    }
}

// ─── Frame Codec ─────────────────────────────────────────────────────────

function encode_frame(int $msgType, string $payload): string
{
    $header = pack('CCCC', MAGIC, VERSION, $msgType, 0x00);
    $header .= pack('N', strlen($payload));
    return $header . $payload;
}

/**
 * @return array{0:int,1:string} [msgType, payload]
 */
function decode_frame(string $data): array
{
    $len = strlen($data);
    if ($len < HEADER_LEN) {
        throw new \InvalidArgumentException("frame too short: $len bytes");
    }
    $magic = ord($data[0]);
    if ($magic !== MAGIC) {
        throw new \InvalidArgumentException(sprintf("bad magic: 0x%02x", $magic));
    }
    $version = ord($data[1]);
    if ($version !== VERSION) {
        throw new \InvalidArgumentException(sprintf("bad version: 0x%02x", $version));
    }
    $msgType = ord($data[2]);
    $payloadLen = unpack('N', substr($data, 4, 4))[1];
    if ($payloadLen !== $len - HEADER_LEN) {
        throw new \InvalidArgumentException("payload length mismatch: declared $payloadLen, available " . ($len - HEADER_LEN));
    }
    return [$msgType, substr($data, HEADER_LEN, $payloadLen)];
}

// ─── Message Class ───────────────────────────────────────────────────────

class Message
{
    public int $type;
    public ?string $clientLanguage = null;
    public ?string $serverLanguage = null;
    public ?int $echoId = null;
    public ?string $echoMeta = null;
    public ?string $echoData = null;
    public ?int $echoStatus = null;
    /** @var array<int,array{idx:int,type:int,kv:array<string,string>}> */
    public ?array $echoResults = null;
    public ?string $osName = null;
    public ?string $osVersion = null;
    public ?string $osRelease = null;
    public ?string $osArch = null;
    public ?string $kissLanguage = null;
    public ?string $kissEncoding = null;
    public ?string $kissTimeZone = null;
    public ?int $timestampMs = null;
    public ?string $iso8601 = null;
    public ?int $randomId = null;
    public ?int $randomNumber = null;
    public ?string $hashHex = null;
    public ?string $disconnectReason = null;
    public ?int $errorCode = null;
    public ?string $errorMessage = null;

    public function __construct(int $type)
    {
        $this->type = $type;
    }

    public function encode(): string
    {
        $w = new ByteWriter();
        switch ($this->type) {
            case MSG_HELLO:
                $w->writeString($this->clientLanguage ?? '');
                break;
            case MSG_BONJOUR:
                $w->writeString($this->serverLanguage ?? '');
                break;
            case MSG_ECHO_REQUEST:
                $w->writeI64($this->echoId ?? 0);
                $w->writeString($this->echoMeta ?? '');
                $w->writeString($this->echoData ?? '');
                break;
            case MSG_ECHO_RESPONSE:
                $w->writeI32($this->echoStatus ?? 200);
                $w->writeU32(count($this->echoResults ?? []));
                foreach ($this->echoResults ?? [] as $r) {
                    $w->writeI64($r['idx']);
                    $w->writeU8($r['type']);
                    $w->writeKv($r['kv']);
                }
                break;
            case MSG_KISS_REQUEST:
                $w->writeString($this->osName ?? '');
                $w->writeString($this->osVersion ?? '');
                $w->writeString($this->osRelease ?? '');
                $w->writeString($this->osArch ?? '');
                break;
            case MSG_KISS_RESPONSE:
                $w->writeString($this->kissLanguage ?? '');
                $w->writeString($this->kissEncoding ?? '');
                $w->writeString($this->kissTimeZone ?? '');
                break;
            case MSG_PING:
                $w->writeI64($this->timestampMs ?? 0);
                break;
            case MSG_PONG:
                $w->writeI64($this->timestampMs ?? 0);
                break;
            case MSG_TIME_NOTIFICATION:
                $w->writeI64($this->timestampMs ?? 0);
                $w->writeString($this->iso8601 ?? '');
                break;
            case MSG_RANDOM_NUMBER:
                $w->writeI64($this->randomId ?? 0);
                $w->writeI64($this->randomNumber ?? 0);
                break;
            case MSG_HASH_RESPONSE:
                $w->writeI64($this->randomId ?? 0);
                $w->writeString($this->hashHex ?? '');
                break;
            case MSG_DISCONNECT:
                $w->writeString($this->disconnectReason ?? '');
                break;
            case MSG_ERROR:
                $w->writeI32($this->errorCode ?? 0);
                $w->writeString($this->errorMessage ?? '');
                break;
            default:
                throw new \InvalidArgumentException("unknown message type: 0x" . dechex($this->type));
        }
        return encode_frame($this->type, $w->data());
    }
}

function decode_message(string $data): Message
{
    [$msgType, $payload] = decode_frame($data);
    $r = new ByteReader($payload);
    $msg = new Message($msgType);

    switch ($msgType) {
        case MSG_HELLO:
            $msg->clientLanguage = $r->readString();
            break;
        case MSG_BONJOUR:
            $msg->serverLanguage = $r->readString();
            break;
        case MSG_ECHO_REQUEST:
            $msg->echoId = $r->readI64();
            $msg->echoMeta = $r->readString();
            $msg->echoData = $r->readString();
            break;
        case MSG_ECHO_RESPONSE:
            $msg->echoStatus = $r->readI32();
            $count = $r->readU32();
            if ($count > intdiv($r->remaining(), 13)) {
                throw new \InvalidArgumentException("result count $count exceeds remaining payload");
            }
            $msg->echoResults = [];
            for ($i = 0; $i < $count; $i++) {
                $idx = $r->readI64();
                $type = $r->readU8();
                $kv = $r->readKv();
                $msg->echoResults[] = ['idx' => $idx, 'type' => $type, 'kv' => $kv];
            }
            break;
        case MSG_KISS_REQUEST:
            $msg->osName = $r->readString();
            $msg->osVersion = $r->readString();
            $msg->osRelease = $r->readString();
            $msg->osArch = $r->readString();
            break;
        case MSG_KISS_RESPONSE:
            $msg->kissLanguage = $r->readString();
            $msg->kissEncoding = $r->readString();
            $msg->kissTimeZone = $r->readString();
            break;
        case MSG_PING:
            $msg->timestampMs = $r->readI64();
            break;
        case MSG_PONG:
            $msg->timestampMs = $r->readI64();
            break;
        case MSG_TIME_NOTIFICATION:
            $msg->timestampMs = $r->readI64();
            $msg->iso8601 = $r->readString();
            break;
        case MSG_RANDOM_NUMBER:
            $msg->randomId = $r->readI64();
            $msg->randomNumber = $r->readI64();
            break;
        case MSG_HASH_RESPONSE:
            $msg->randomId = $r->readI64();
            $msg->hashHex = $r->readString();
            break;
        case MSG_DISCONNECT:
            $msg->disconnectReason = $r->readString();
            break;
        case MSG_ERROR:
            $msg->errorCode = $r->readI32();
            $msg->errorMessage = $r->readString();
            break;
        default:
            throw new \InvalidArgumentException("unknown message type: 0x" . dechex($msgType));
    }

    return $msg;
}

// ─── Factory Helpers ─────────────────────────────────────────────────────

function hello(string $clientLanguage): Message
{
    $m = new Message(MSG_HELLO);
    $m->clientLanguage = $clientLanguage;
    return $m;
}

function bonjour(string $serverLanguage): Message
{
    $m = new Message(MSG_BONJOUR);
    $m->serverLanguage = $serverLanguage;
    return $m;
}

function ping(int $timestampMs): Message
{
    $m = new Message(MSG_PING);
    $m->timestampMs = $timestampMs;
    return $m;
}

function pong(int $timestampMs): Message
{
    $m = new Message(MSG_PONG);
    $m->timestampMs = $timestampMs;
    return $m;
}

function timeNotif(int $timestampMs, string $iso8601): Message
{
    $m = new Message(MSG_TIME_NOTIFICATION);
    $m->timestampMs = $timestampMs;
    $m->iso8601 = $iso8601;
    return $m;
}

function randomNumberMsg(int $id, int $number): Message
{
    $m = new Message(MSG_RANDOM_NUMBER);
    $m->randomId = $id;
    $m->randomNumber = $number;
    return $m;
}

function hashResponseMsg(int $id, string $hashHex): Message
{
    $m = new Message(MSG_HASH_RESPONSE);
    $m->randomId = $id;
    $m->hashHex = $hashHex;
    return $m;
}

function kissRequest(string $osName, string $osVersion, string $osRelease, string $osArch): Message
{
    $m = new Message(MSG_KISS_REQUEST);
    $m->osName = $osName;
    $m->osVersion = $osVersion;
    $m->osRelease = $osRelease;
    $m->osArch = $osArch;
    return $m;
}

function kissResponse(string $language, string $encoding, string $timeZone): Message
{
    $m = new Message(MSG_KISS_RESPONSE);
    $m->kissLanguage = $language;
    $m->kissEncoding = $encoding;
    $m->kissTimeZone = $timeZone;
    return $m;
}

function disconnectMsg(string $reason): Message
{
    $m = new Message(MSG_DISCONNECT);
    $m->disconnectReason = $reason;
    return $m;
}

function errorMsg(int $code, string $message): Message
{
    $m = new Message(MSG_ERROR);
    $m->errorCode = $code;
    $m->errorMessage = $message;
    return $m;
}

// ─── Utility ─────────────────────────────────────────────────────────────

function now_ms(): int
{
    return (int)(microtime(true) * 1000);
}

function now_iso(): string
{
    return gmdate('Y-m-d\TH:i:s\Z');
}

function hash_number(int $num): string
{
    return substr(hash('sha256', (string)$num), 0, 10);
}

function log_msg(string $name, string $msg): void
{
    $ts = date('Y-m-d H:i:s');
    echo "[$ts] [INFO] [$name] $msg\n";
}
